products/swift-Ninja/NinjaManifest.swift

```diff
@@ -0,0 +1,118 @@
+    /// The description of this rule, if present.
+    public let description: String?
+
+    public init(
+        command: String? = nil,
+        description: String? = nil
+    ) {
+        self.command = command
+        self.description = description
+    }
+}
+
+/// An individual command in a Ninja manifest.
+public struct Command: Codable, Equatable {
+    /// The name of the rule which produced this command.
+    public let rule: String
+
+    /// The list of inputs to the command.
+    public let inputs: [String]
+
+    /// The list of outputs of the command.
+    public let outputs: [String]
+
+    /// The command itself (a string to run with "/bin/sh -c").
+    public let command: String
+
+    /// The description of the string, or empty.
+    public let description: String
+
+    public init(
+        rule: String,
+        inputs: [String],
+        outputs: [String],
+        command: String,
+        description: String
+    ) {
+        self.rule = rule
+        self.inputs = inputs
+        self.outputs = outputs
+        self.command = command
+        self.description = description
+    }
+}
+
+public class NinjaManifest: Codable {
+    public enum Error: Swift.Error {
+        case unableToLocateBinary
+    }
+
+    /// The name of the source manifest file.
+    public let filename: String
+
+    /// The dictionary of build rules.
+    public let rules: [String: Rule]
+
+    /// The list of build commands.
+    public let commands: [Command]
+
+    private static func findBinaryPath() throws -> URL {
+        guard let exePath = Bundle.main.executablePath else {
+            throw Error.unableToLocateBinary
+        }
+
+        // Return the FooBar.xctest name if we're running under xctest.
+        if exePath.hasSuffix("xctest") {
+            for bundle in Bundle.allBundles {
+                if bundle.bundlePath.hasSuffix(".xctest") {
+                    return URL(fileURLWithPath: bundle.bundlePath)
+                }
+            }
+        }
+
+        return URL(fileURLWithPath: exePath)
+    }
+    
+    private static func findLLBuildBinary() throws -> URL {
+        return try findBinaryPath().deletingLastPathComponent().appendingPathComponent("llbuild")
+    }
+    
+    public init(path: String) throws {
+        // For now, we depend on the `llbuild` binary to be adjust to the
+        // package using this logic, rather than building APIs to llbuild's
+        // Ninja layer. This will eventually need to get cleaned up.
+        let p = Process()
+        p.launchPath = try Self.findLLBuildBinary().path
+        p.arguments = ["ninja", "load-manifest", "--json", path]
+        let pipe = Pipe()
+        p.standardOutput = pipe
+        p.launch()
+        var data: Data!
+        let sema = DispatchSemaphore(value: 0)
+        DispatchQueue.global().async {
+            data = pipe.fileHandleForReading.readDataToEndOfFile()
+            sema.signal()
+        }
+        p.waitUntilExit()
+        sema.wait()
```

Comment:
Are we certain this version is safe against deadlocks? Lily was helping me with a similar issue earlier...

/CC @millenomi

Reply:
I'd be very interested to know how this could deadlock.  The calling thread is blocked until process exit.  The pipe reader will continue to read until the pipe is closed, aka process exit.  It isn't waiting on the caller thread for anything.   The caller thread doesn't wait on the semaphore until after the process has exited, but at that point we should reasonably expect that the pipe read will be getting the EOF soon, if it hasn't already.

Reply:
Exactly, certainly without the async it can easily deadlock